model_analyzer/analyzer.py

```diff
@@ -0,0 +1,227 @@
+        self.tables["Server Only:"].add_row(output_row)
+        dcgm_monitor.destroy()
+
+    def profile_model(self, run_config, write_perf_output=False):
+        """
+        Runs DCGMMonitor while running perf_analyzer
+        with a specific set of arguments. This will
+        profile model inferencing.
+
+        Parameters
+        ----------
+        run_config : dict
+            The keys are arguments to perf_analyzer
+            The values are their values
+        write_perf_output : boolean
+            Whether or not to write the output from
+            perf_analyzer to the output stream/file
+        """
+
+        dcgm_monitor = DCGMMonitor(self.monitoring_interval, self.dcgm_tags)
+        perf_analyzer = PerfAnalyzer(
+            path=self.perf_analyzer_path,
+            config=self._create_perf_config(run_config))
+
+        # Get metrics for model inference and write perf_output
+
+        model_metrics = self._profile(perf_analyzer=perf_analyzer,
+                                      dcgm_monitor=dcgm_monitor)
+        if write_perf_output:
+            self.writer.write(perf_analyzer.output() + '\n')
+
+        output_row = [
+            run_config['model-name'], run_config['batch-size'],
+            run_config['concurrency-range']
+        ]
+        output_row += [
+            model_metrics[metric] for metric in self.monitoring_metrics
+        ]
+        self.tables["Models:"].add_row(output_row)
+
+        dcgm_monitor.destroy()
+
+    def write_results(self, column_width, column_separator):
+        """
+        Writes the tables using the writer with the given
+        column specifications.
+
+        Parameters
+        ----------
+        column_width : int
+            The width of each column of the written table
+        column_separator : str
+            The string that will be inserted between each column
+            of the table
+        """
+
+        for table_name in self.tables:
+            self.writer.write('\n'.join([
+                table_name,
+                self.tables[table_name].to_formatted_string(column_width=28,
+                                                            separator=' '),
+                "\n"
+            ]))
+
+    def _profile(self, perf_analyzer, dcgm_monitor):
+        """
+        Utility function that runs the perf_analyzer
+        and DCGMMonitor once.
+
+        Raises
+        ------
+        TritonModelAnalyzerException
+            if path to perf_analyzer binary could 
+            not be found.
+        """
+
+        # Start monitors and run perf_analyzer
+        dcgm_monitor.start_recording_metrics()
+        if perf_analyzer:
+            try:
+                perf_records = perf_analyzer.run(self.perf_tags)
+            except FileNotFoundError as e:
+                raise TritonModelAnalyzerException(
+                    f"perf_analyzer binary not found : {e}")
+        else:
+            perf_records = []
+            time.sleep(0.1)
```

Comment:
Why should we sleep here?

Reply:
I ran into an issue where if I try to run 
``` 
dcgm_monitor.start_recording_metrics()
dcgm_monitor.stop_recording_metrics()
```
No metrics are collected because the monitoring thread is inactive when `stop_recording_metrics` is called. This triggers a TritonModelAnalyzerException [here](https://github.com/triton-inference-server/model_analyzer/blob/6b4f68469e198ce4959da9433c82929ca6903092/model_analyzer/monitor/monitor.py#L108)

Reply:
I'll look into it. It looks like a concurrency issue.

Reply:
Sounds good.